cupy/cuda/memory.pyx

```diff
@@ -557,7 +557,12 @@ cpdef set_allocator(allocator=None):
     global _current_allocator
     if allocator is None:
         allocator = _malloc
-    _current_allocator = allocator
+
+    is_local = getattr(_thread_local, 'allocator', None)
+    if is_local is not None:
+        _thread_local.allocator = allocator
```

Comment:
I feel this behavior is not intuitive. How about raising an error in such a case?

Reply:
That's the other alternative, I asked about it earlier today :)

Reply:
@asi1024 Why do you feel it's not intuitive?
Error is fine, but this behavior looks fine to me, too.

Reply:
OK, I agree it's not great in a situation like this:
```
def setup_global_state():
    set_allocator(...)

def main():
    setup_global_state()
    ... # main logic

with using_allocator(...):
    main()

```

Reply:
Pushed an alternative that throws an exception in such case